File: swift/Sources/SecRandom.m
// Copyright © 2017-2018 Trust.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

@import Security;

int random32(uint32_t *result) {
    if (SecRandomCopyBytes(kSecRandomDefault, sizeof(uint32_t), &result) != errSecSuccess) {
        // failed to generate random number
        return -1;
    }
    return 0;
}

int random_buffer(uint8_t *buf, size_t len) {
    if (SecRandomCopyBytes(kSecRandomDefault, len, buf) != errSecSuccess) {
        return -1;
    }
    return 0;
}

File: swift/Sources/SecRandom.swift
// Copyright © 2017-2020 Trust Wallet.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

import Security

@_cdecl("random32")
public func random32() -> UInt32 {
    var bytes = [UInt8](repeating: 0, count: 32)
    guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
        // failed to generate random number
        abort()
    }

    return bytes.withUnsafeBytes { ptr -> UInt32 in
        ptr.baseAddress!.assumingMemoryBound(to: UInt32.self).pointee.littleEndian
    }
}

@_cdecl("random_buffer")
public func random_buffer(buf: UnsafeMutableRawPointer, len: Int) {
    guard SecRandomCopyBytes(kSecRandomDefault, len, buf) == errSecSuccess else {
        // failed to generate random number
        abort()
    }
}
